Add Create an NFT Collection using LSP8 tutorial (#715)
lukso-network/docs: docs/learn/smart-contract-developers/create-nft-collection.md

---
sidebar_label: '🗃 Create an NFT Collection'
sidebar_position: 4
---

# Create an NFT Collection Using LSP8

This tutorial goes over creating a collection of collectibles, where each item is unique and has its own design.

:::info
⌨️ The full code of this example can be found in 👾 [here](https://github.com/CJ42/LUKSO-Hardhat-template).
:::

## Setup

Make sure you have the following dependencies installed:

- [`ethers.js`](https://github.com/ethers-io/ethers.js/) (alternatively you can use [`web3.js`](https://github.com/web3/web3.js))
- [`@lukso/lsp-smart-contracts`](https://github.com/lukso-network/lsp-smart-contracts/)

> 

Reviewer: @biancabuzea200 can you stay consistent and make it look like the other guides with tabs please?

<img width="263" alt="image" src="https://github.com/lukso-network/docs/assets/477945/edfbab6a-e147-4e3a-8567-72c2f8a98232">

https://docs.lukso.tech/learn/dapp-developer/transfer-lyx#setup

And adapt the code below where you use etherjs to add a web3 version (using tab)

Link this:
https://github.com/lukso-network/docs/blob/main/docs/learn/dapp-developer/transfer-lyx.md?plain=1#L56

with tab group id:
https://docusaurus.io/docs/markdown-features/tabs#syncing-tab-choices

> 

Reviewer: This was marked as solved but I don't see it solved. @biancabuzea200

## Create the Smart Contracts

When creating digital asset smart contracts on LUKSO, you will need to specify the token type and the following data keys in the ERC725Y storage. There are three types that you can define:

> - 0 = Token
> - 1 = NFT
> - 2 = Collection

Reviewer 2: Does the number have a meaning? If so, I would suggest putting it in code style.

```suggestion
- `0` = Token
- `1` = NFT
- `2` = Collection
```

```solidity title="contracts/TokenTypes.sol"
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.4;

bytes32 constant _LSP4_TOKEN_TYPE_DATA_KEY = 0xe0261fa95db2eb3b5439bd033cda66d56b96f92f243a8228fd87550ed7bdfdb3;

enum TokenType {
    TOKEN,
    NFT,
    COLLECTION
}

```

> You will also create a custom [LSP8 Identfiable Digital Asset Collection](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md) that extends [LSP8Mintable](../../contracts/contracts/LSP8IdentifiableDigitalAsset/presets/LSP8Mintable.md).

Reviewer 2: ```suggestion
After defining the type of the asset and its 🗂️ [ERC725 data key](../../standards/lsp-background/erc725.md#erc725y-generic-data-keyvalue-store) you can create a custom 🌄 [LSP8 Identfiable Digital Asset Collection](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md) that extends [LSP8Mintable](../../contracts/contracts/LSP8IdentifiableDigitalAsset/presets/LSP8Mintable.md) so that new assets can be created within the smart contract.
```

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.4;

// modules
import {
    LSP8Mintable
} from "@lukso/lsp-smart-contracts/contracts/LSP8IdentifiableDigitalAsset/presets/LSP8Mintable.sol";

// constants
import {
    _LSP8_TOKENID_TYPE_NUMBER
} from "@lukso/lsp-smart-contracts/contracts/LSP8IdentifiableDigitalAsset/LSP8Constants.sol";
import {_LSP4_TOKEN_TYPE_DATA_KEY, TokenType} from "../TokenTypes.sol";

contract BasicNFTCollection is LSP8Mintable {
    constructor(
        string memory nftCollectionName,
        string memory nftCollectionSymbol,
        address contractOwner
    )
        LSP8Mintable(
            nftCollectionName,        // NFT collection name
            nftCollectionSymbol,      // NFT collection symbol
            contractOwner,            // owner of the NFT contract (the address that controls it, sets metadata, can transfer the ownership of the contract)
            _LSP8_TOKENID_TYPE_NUMBER // type of NFT/ tokenIds
        )
    {
        // set token type
        _setData(_LSP4_TOKEN_TYPE_DATA_KEY, abi.encode(TokenType.COLLECTION));
    }
}
```

## Deploy our NFT Collection Contract on the LUKSO Testnet

> 

Reviewer 2: I would suggest adding a sentence here to introduce the people into the next section, maybe even put a "info" box for hardhat for which network it can be deployed.

Explain why `import {BasicNFTCollection, BasicNFTCollection__factory} from "../typechain-types";` is needed and where users are able to specify their network for the deployment (maybe just a notice: `it will automatically use your default network setup`)

```js
import { ethers } from "hardhat";

import {BasicNFTCollection, BasicNFTCollection__factory} from "../typechain-types";

async function deployLSP8Collection() {
  const accounts = await ethers.getSigners();
  const deployer = accounts[0];

  const nftCollection: BasicNFTCollection = await new BasicNFTCollection__factory(deployer).deploy(
    "NFT Collection Name", // collection name
    "NFT",                 // collection symbol
    deployer.address
  );

}

deployLSP8Collection().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

```

> ## Congratulations 🥳
> 
> You have deployed your first LSP8 NFT Collection on the LUKSO Testnet.

Reviewer 2: I know we used such sections in the old guides, but i would remove it completely. it adds no context. 

I would exchange it with more context:

```suggestion
## Retrieve the Digital Asset

--> Information on how the asset that was deployed can be seen/checked/retrieved 
- (backend) in the block explorer 
- (web) on universal profiles.cloud (if testnet)
- (local) using the "fetchData" function (get profile data guide...but instead of UP, use asset address)
```

---
sidebar_label: '🗃 Create an NFT Collection'
sidebar_position: 4
---

# Create an NFT Collection Using LSP8

This tutorial will teach you how to create a collection of unique digital assets.

:::info
⌨️ The full code of this example can be found in the 👾 [lukso-hardhat-template](https://github.com/CJ42/LUKSO-Hardhat-template).
:::

## Setup

Make sure you have the following dependencies installed:

- [`ethers.js`](https://github.com/ethers-io/ethers.js/) (alternatively you can use [`web3.js`](https://github.com/web3/web3.js))
- [`@lukso/lsp-smart-contracts`](https://github.com/lukso-network/lsp-smart-contracts/)

\```bash
npm install ethers @lukso/lsp-smart-contracts
\```

## Create the Smart Contracts

When creating digital asset smart contracts on LUKSO, you will need to specify the token type and the following data keys in the ERC725Y storage. There are three types that you can define:

- `0` = Token
- `1` = NFT
- `2` = Collection

```solidity title="contracts/TokenTypes.sol"
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.4;

> bytes32 constant _LSP4_TOKEN_TYPE_DATA_KEY = 0xe0261fa95db2eb3b5439bd033cda66d56b96f92f243a8228fd87550ed7bdfdb3;

Author: Waiting for additional info from @CJ42 about how this will be implemented based on this https://github.com/lukso-network/LIPs/pull/250

```solidity title="contracts/MyNFTCollection.sol"

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.4;

// modules
import {
    LSP8Mintable
} from "@lukso/lsp-smart-contracts/contracts/LSP8IdentifiableDigitalAsset/presets/LSP8Mintable.sol";

// constants
import {
    _LSP8_TOKENID_TYPE_NUMBER
} from "@lukso/lsp-smart-contracts/contracts/LSP8IdentifiableDigitalAsset/LSP8Constants.sol";
import {_LSP4_TOKEN_TYPE_DATA_KEY, TokenType} from "../TokenTypes.sol";

contract BasicNFTCollection is LSP8Mintable {
    constructor(
        string memory nftCollectionName,
        string memory nftCollectionSymbol,
        address contractOwner
    )
        LSP8Mintable(
            nftCollectionName,        // NFT collection name
            nftCollectionSymbol,      // NFT collection symbol
            contractOwner,            // owner of the NFT contract (the address that controls it, sets metadata, can transfer the ownership of the contract)
            _LSP8_TOKENID_TYPE_NUMBER // type of NFT/ tokenIds
        )
    {
        // set token type
        _setData(_LSP4_TOKEN_TYPE_DATA_KEY, abi.encode(TokenType.COLLECTION));
    }
}
```

> ## Deploy our NFT Collection Contract on the LUKSO Testnet

Reviewer: Our -> your

```js
import { ethers } from "hardhat";

import {BasicNFTCollection, BasicNFTCollection__factory} from "../typechain-types";

async function deployLSP8Collection() {
  const accounts = await ethers.getSigners();
  const deployer = accounts[0];

  const nftCollection: BasicNFTCollection = await new BasicNFTCollection__factory(deployer).deploy(
    "NFT Collection Name", // collection name
    "NFT",                 // collection symbol
    deployer.address
  );

}

deployLSP8Collection().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

```

## Congratulations 🥳

> You have deployed your first LSP8 NFT Collection on the LUKSO Testnet.

Reviewer: Suggest a place where the developer can check the contract. At the very least a link to the execution explorer. But maybe a link to the wallet or other NFT marketplace could be useful too.

---
sidebar_label: '🗃 Create an NFT Collection'
sidebar_position: 4
---

# Create an NFT Collection Using LSP8

> This tutorial will explore how to create a collection of unique digital assets.

Reviewer: Don't forget to add links to other part so readers can explore more.

```suggestion
This tutorial will explore how to create a collection of unique [digital assets](../{..}/standards/tokens/LSP8-Identifiable-Digital-Asset).
```

please adjust the link accordingly

> :::info
> ⌨️ The full code of this example can be found in the 👾 [lukso-hardhat-template](https://github.com/CJ42/LUKSO-Hardhat-template).
> :::

Reviewer: Can we stay consistent?

Why this info has no blank lines why the note below has blank line?

```suggestion
:::info

⌨️ The full code of this example can be found in the 👾 [lukso-hardhat-template](https://github.com/CJ42/LUKSO-Hardhat-template).

:::
```

Reason:
https://docusaurus.io/docs/markdown-features/admonitions#usage-with-prettier

:::note

> This guide builds on top of a Hartdhat project using TypeScript as described in the [Getting Started section](../smart-contract-developers/getting-started.md).

Reviewer: typo Hartdhat

- `0` = Token
- `1` = NFT
- `2` = Collection

```solidity title="contracts/TokenTypes.sol"
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.4;

bytes32 constant _LSP4_TOKEN_TYPE_DATA_KEY = 0xe0261fa95db2eb3b5439bd033cda66d56b96f92f243a8228fd87550ed7bdfdb3;

enum TokenType {
    TOKEN,
    NFT,
    COLLECTION
}
```

After defining the type of the asset and its 🗂️ [ERC725 data key](../../standards/lsp-background/erc725.md#erc725y-generic-data-keyvalue-store) you can create a custom 🌄 [LSP8 Identfiable Digital Asset Collection](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md) that extends [LSP8Mintable](../../contracts/contracts/LSP8IdentifiableDigitalAsset/presets/LSP8Mintable.md) so that new assets can be created within the smart contract.

```solidity title="contracts/MyNFTCollection.sol"

> 

Reviewer: why blank line here?

Can you remove please?

            nftCollectionSymbol,      // NFT collection symbol
            contractOwner,            // owner of the NFT contract (the address that controls it, sets metadata, can transfer the ownership of the contract)
            _LSP8_TOKENID_TYPE_NUMBER // type of NFT/ tokenIds
        )
    {
        // set token type
        _setData(_LSP4_TOKEN_TYPE_DATA_KEY, abi.encode(TokenType.COLLECTION));
    }
}
```

## Deploy your NFT Collection Contract on the LUKSO Testnet

:::info

By default, the deployment will be to your local network. If you want to deploy to the LUKSO Testnet, you will need to add the LUKSO Testnet network in your `hardhat.config.ts`.

:::

```js
function getTestnetChainConfig(): NetworkUserConfig {
  const config: NetworkUserConfig = {
    url: 'https://rpc.testnet.lukso.network',
    chainId: 4201,
  };

  if (process.env.PRIVATE_KEY !== undefined) {
    config['accounts'] = [process.env.PRIVATE_KEY];
  }

  return config;
}
```

Also add a definition for the testnet in the `HardhatUserConfig`:

```js
const config: HardhatUserConfig = {
  solidity: {
    version: '0.8.19',
    settings: {
      optimizer: {
        enabled: true,
        runs: 200,

>       },

Reviewer: I think this part should go in the "Getting Started" section as it applies to all the other tutorials and it is not specific to "deploy lsp8 collection"

    },
  },
  typechain: {
    outDir: 'typechain-types',
    target: 'ethers-v6',
  },
  networks: {
    luksoTestnet: getTestnetChainConfig(),
  },
};
```

Next you define the deployment script:

```js
import { ethers } from "hardhat";

import {BasicNFTCollection, BasicNFTCollection__factory} from "../typechain-types";

async function deployLSP8Collection() {
  const accounts = await ethers.getSigners();
  const deployer = accounts[0];

  const nftCollection: BasicNFTCollection = await new BasicNFTCollection__factory(deployer).deploy(
    "NFT Collection Name", // collection name
    "NFT",                 // collection symbol
    deployer.address
  );

}

deployLSP8Collection().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

```

## Check Your NFT Collection

> You can now check your NFT collection using the [execution block explorer](https://explorer.execution.testnet.lukso.network/).

Reviewer: Can you explain which address to check?

Or make the script in a way that it outputs the deployment contract address?
Maybe some screenshots?

---
sidebar_label: '🗃 Create an NFT Collection'
sidebar_position: 4
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Create an NFT Collection Using LSP8

This tutorial will explore how to create a collection of unique [digital assets](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md).

:::info

⌨️ The full code of this example can be found in the 👾 [lukso-hardhat-template](https://github.com/CJ42/LUKSO-Hardhat-template).

:::

:::note

This guide builds on top of a Hardhat project using TypeScript as described in the [Getting Started](../smart-contract-developers/getting-started.md) section.

:::

## Setup

Make sure you have the following dependencies installed:

> - [`ethers.js`](https://github.com/ethers-io/ethers.js/) (alternatively you can use [`web3.js`](https://github.com/web3/web3.js))

Reviewer 3: This is important to clarify.

```suggestion
- [`ethers.js v6`](https://github.com/ethers-io/ethers.js/) (alternatively you can use [`web3.js`](https://github.com/web3/web3.js))
```

Author: Thanks. Updated. What is incompatible if v6 is not used?

Reviewer 3: Oh wait, actually it looks like only the `getSigner` from ethers is used. Which is the same between v5 and v6. So I think it should be ok for this code snippet if whether v5 or v6 is used. My bad 🙈

Author: no worries, thank you for looking into it

- [`@lukso/lsp-smart-contracts`](https://github.com/lukso-network/lsp-smart-contracts/)

```bash
npm install ethers @lukso/lsp-smart-contracts
```

## Create the Smart Contracts

> When creating smart contracts representing digital assets on LUKSO, you will need to specify the token type and data keys for the 📄 [LSP4 Digital Asset Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata) that will be stored in the 🗂️ [ERC725Y](../../standards/lsp-background/erc725.md#erc725y-generic-data-keyvalue-store) storage of the Universal Profile. There are three different token types:

Reviewer 3: These data keys are stored under the data key-value store of the digital asset.

```suggestion
When creating smart contracts representing digital assets on LUKSO, you will need to specify the token type and data keys for the 📄 [LSP4 Digital Asset Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata) that will be stored in the 🗂️ [ERC725Y](../../standards/lsp-background/erc725.md#erc725y-generic-data-keyvalue-store) storage of the Digital Asset. There are three different token types:
```

- `0` = Token
- `1` = NFT
- `2` = Collection

```solidity title="contracts/TokenTypes.sol"
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.4;

bytes32 constant _LSP4_TOKEN_TYPE_DATA_KEY = 0xe0261fa95db2eb3b5439bd033cda66d56b96f92f243a8228fd87550ed7bdfdb3; // kecca256 hash of the word `LSP4TokenType`

enum TokenType {
    TOKEN,
    NFT,
    COLLECTION
}
```

After defining the type of the asset and its 🗂️ [ERC725 data key](../../standards/lsp-background/erc725.md#erc725y-generic-data-keyvalue-store) you can create a custom 🌄 [LSP8 Identfiable Digital Asset Collection](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md) that extends [LSP8Mintable](../../contracts/contracts/LSP8IdentifiableDigitalAsset/presets/LSP8Mintable.md) so that new assets can be created within the smart contract.

> ```solidity title="contracts/MyNFTCollection.sol"

Reviewer: This file is not in the repo that you link above:

<img width="324" alt="image" src="https://github.com/lukso-network/docs/assets/477945/4b18d7bd-8648-42bd-8499-e804b59ba081">

<img width="604" alt="image" src="https://github.com/lukso-network/docs/assets/477945/c5570396-13f7-4b0c-9ec4-33d7bd767063">

??

// modules
import {
    LSP8Mintable
} from "@lukso/lsp-smart-contracts/contracts/LSP8IdentifiableDigitalAsset/presets/LSP8Mintable.sol";

// constants
import {
    _LSP8_TOKENID_TYPE_NUMBER
} from "@lukso/lsp-smart-contracts/contracts/LSP8IdentifiableDigitalAsset/LSP8Constants.sol";
import {_LSP4_TOKEN_TYPE_DATA_KEY, TokenType} from "../TokenTypes.sol";

contract BasicNFTCollection is LSP8Mintable {
    constructor(
        string memory nftCollectionName,
        string memory nftCollectionSymbol,
        address contractOwner
    )
        LSP8Mintable(
            nftCollectionName,        // NFT collection name
            nftCollectionSymbol,      // NFT collection symbol
            contractOwner,            // owner of the NFT contract (the address that controls it, sets metadata, can transfer the ownership of the contract)
            _LSP8_TOKENID_TYPE_NUMBER // type of NFT/ tokenIds
        )
    {
        // set token type
        _setData(_LSP4_TOKEN_TYPE_DATA_KEY, abi.encode(TokenType.COLLECTION));
    }
}
```

Next you define the deployment script.

<Tabs groupId="web3-lib">
  <TabItem value="ethersjs" label="ethers.js">

<!-- prettier-ignore-start -->

> ```js

Reviewer 3: ```suggestion
```js title="deployLSP8Collection.ts"
```

> ```js title="deployLSP8Collection.ts"

Reviewer: same this file is nowhere to be found in the related repo

<img width="350" alt="image" src="https://github.com/lukso-network/docs/assets/477945/1eea5a97-4bfb-4be3-888d-6490030c9f3a">

https://github.com/CJ42/LUKSO-Hardhat-template

import { ethers } from "hardhat";
import {BasicNFTCollection, BasicNFTCollection__factory} from "../typechain-types";

async function deployLSP8Collection() {
  const accounts = await ethers.getSigners();
  const deployer = accounts[0];

>   const nftCollection: BasicNFTCollection = await new BasicNFTCollection\_\_factory(deployer).deploy(

Reviewer 3: Not sure if it is normal to have the backslash characters here?

```suggestion
  const nftCollection: BasicNFTCollection = await new BasicNFTCollection__factory(deployer).deploy(
```

      "NFT Collection Name", // collection name
      "NFT", // collection symbol
      deployer.address
   );
  const nftCollectionAddress = await nftCollection.getAddress()
  console.log("NFT Collection deployed to:", nftCollectionAddress)
  console.log("Check the block explorer to see the deployed contract")
}

deployLSP8Collection().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

```
<!-- prettier-ignore-end -->
</TabItem>
<TabItem value="web3js" label="web3.js">

Swap to the latest version of node;

```sh
nvm install node
nvm use node
```

> 

Reviewer: this is not in the context of this article. can you remove?
